shims/spark310/src/main/scala/com/nvidia/spark/rapids/shims/spark310/ParquetCachedBatchSerializer.scala

```diff
@@ -0,0 +1,303 @@
+    bytes += len
+  }
+
+  def getBuffer(): Array[Byte] = {
+    if (buffer == null) {
+      writeBuffers()
+    }
+    buffer
+  }
+
+  def close(): Unit = {
+    writeBuffers()
+  }
+
+  def writeBuffers(): Unit = {
+    // this could be problematic if the buffers are big as their cumulative length could be more
+    // than an Int.MAX_SIZE. We could just have a list of buffers in that case and iterate over them
+    buffer = new Array(sizeInBytes.toInt)
+    val toProcess = offHeapBuffers.dequeueAll(_ => true)
+    try {
+      var offset: Int = 0
+      toProcess.foreach(ops => {
+        val origBuffer = ops._1
+        val len = ops._2.toInt
+        origBuffer.asByteBuffer().get(buffer, offset, len)
+        for (i <- 0  until len) {
+          assert(origBuffer.getByte(i) == buffer(offset + i.toInt))
+        }
+        offset = offset + len
+      })
+    } finally {
+      toProcess.map(_._1).safeClose()
+    }
+  }
+}
+
+private case class CloseableColumnBatchIterator(iter: Iterator[ColumnarBatch]) extends
+  Iterator[ColumnarBatch] {
+  var cb: ColumnarBatch = null
+
+  private def closeCurrentBatch(): Unit = {
+    if (cb != null) {
+      cb.close
+      cb = null
+    }
+  }
+
+  TaskContext.get().addTaskCompletionListener[Unit]((tc: TaskContext) => {
+    closeCurrentBatch()
+  })
+
+  override def hasNext: Boolean = iter.hasNext
+
+  override def next(): ColumnarBatch = {
+    closeCurrentBatch()
+    cb = iter.next()
+    cb
+  }
+}
+
+/**
+ * This class assumes, the data is Columnar and the plugin is on
+ */
+class ParquetCachedBatchSerializer extends CachedBatchSerializer with Arm {
+  override def supportsColumnarInput(schema: Seq[Attribute]): Boolean = true
+
+  override def supportsColumnarOutput(schema: StructType): Boolean = true
+
+  /**
+   * Convert an `RDD[ColumnarBatch]` into an `RDD[CachedBatch]` in preparation for caching the data.
+   * This method uses Parquet Writer on the GPU to write the cached batch
+   * @param input the input `RDD` to be converted.
+   * @param schema the schema of the data being stored.
+   * @param storageLevel where the data will be stored.
+   * @param conf the config for the query.
+   * @return The data converted into a format more suitable for caching.
+   */
+  override def convertColumnarBatchToCachedBatch(input: RDD[ColumnarBatch],
+                                       schema: Seq[Attribute],
+                                       storageLevel: StorageLevel,
+                                       conf: SQLConf): RDD[CachedBatch] = {
+    val parquetCB = input.map(batch => {
+      var gpuCB: ColumnarBatch = null
+      try {
+        // check if the data is already on GPU
+        if (!batch.column(0).isInstanceOf[GpuColumnVector]) {
+          val s = StructType(
+            schema.map(a => StructField(a.name, a.dataType, a.nullable, a.metadata)))
+          gpuCB = new GpuColumnarBatchBuilder(s, batch.numRows(), batch).build(batch.numRows())
+          batch.close()
```

Comment:
Technically if it is on the CPU then it is the responsibility of the producer to close the batch, not the consumer.

This is one of the odd things with spark.  They expect the producer to close the batch, where as we expect the consumer to do it.  They did this so that they could reuse the memory for a batch whenever possible, mostly because it is used in just a limited number of locations and with a very limited size of a batch.